lib/src/entrypoint.dart

```diff
@@ -284,6 +288,122 @@ class Entrypoint {
+  /// from scratch. Otherwise, it will attempt to preserve the versions of all
+  /// previously locked packages.
+  ///
+  /// Shows a report of the changes made relative to the previous lockfile. If
+  /// this is an upgrade or downgrade, all transitive dependencies are shown in
+  /// the report. Otherwise, only dependencies that were changed are shown. If
+  /// [dryRun] is `true`, no physical changes are made.
+  ///
+  /// If [precompile] is `true` (the default), this snapshots dependencies'
+  /// executables.
+  ///
+  /// Updates pubspec.yaml, [lockFile] and [packageRoot] accordingly.
+  Future addAndAcquireDependencies(SolveType type, List<String> packages,
+      {List<String> useLatest,
+      bool development = false,
+      bool dryRun = false,
+      bool precompile = false}) async {
+    if (root.pubspecPath == null) {
+      throw FileException(
+          // Make the package dir absolute because for the entrypoint it'll just
+          // be ".", which may be confusing.
+          'Could not find a file named "pubspec.yaml" in '
+          '"${canonicalize('.')}".',
+          root.pubspecPath);
+    }
+
+    final dependencyKey = development ? 'dev_dependencies' : 'dependencies';
+
+    final yamlEditor = YamlEditor(readTextFile(pubspecPath));
+
+    for (var package in packages) {
+      yamlEditor.assign([dependencyKey, package], 'any');
+    }
+
+    final pubspec = Pubspec.parse(yamlEditor.toString(), cache.sources,
+        includeDefaultSdkConstraint: false, // ?
+        location: p.toUri(root.pubspecPath));
+
+    final package = Package(pubspec, '.');
+
+    final result = await log.progress(
+      'Resolving dependencies',
+      () => resolveVersions(
+        type,
+        cache,
+        package,
+        lockFile: lockFile,
+        useLatest: useLatest,
+      ),
+    );
+
+    // Log once about all overridden packages.
+    if (warnAboutPreReleaseSdkOverrides && result.pubspecs != null) {
+      var overriddenPackages = (result.pubspecs.values
+              .where((pubspec) => pubspec.dartSdkWasOverridden)
+              .map((pubspec) => pubspec.name)
+              .toList()
+                ..sort())
+          .join(', ');
+      if (overriddenPackages.isNotEmpty) {
+        log.message(log.yellow(
+            'Overriding the upper bound Dart SDK constraint to <=${sdk.version} '
+            'for the following packages:\n\n$overriddenPackages\n\n'
+            'To disable this you can set the PUB_ALLOW_PRERELEASE_SDK system '
+            'environment variable to `false`, or you can silence this message '
+            'by setting it to `quiet`.'));
+      }
+    }
+
+    result.showReport(type);
+    final finalPackages = <String, Version>{};
+    for (var package in result.packages) {
+      finalPackages[package.name] = package.version;
+    }
+
+    for (var package in packages) {
+      yamlEditor.assign([dependencyKey, package], '^${finalPackages[package]}');
```

Comment:
What if I have a dependency override on the package I'm adding?

What if the package I'm adding is from a git source... let's at-least we should check that the source is a `HostedSource`

Reply:
What if it was a dev dependency before..

Reply:
I think if you had something like:
```
dependencies:
  retry: ^3.0.0

dev-dependencies:
  retry: '< 2.0.0'
```

`pub get` works, and you'll be using `retry 1.0.0` locally, so in this case I don't think the dependency override matters?


With regards to it being a dev dependency before, I think the behaviour we should go with is to remove the dev dependency, because if there's a normal dependency then the dev dependency is unnecessary. Arguably there might be a use case where the dev dependency needs to be a different version than the normal dependency...not too sure how often that will come into play though

Reply:
I just tested with yarn, and if it's already present in dependencies and you want to add it to dev-dependencies (or vice versa), it's a no op with a printed warning - maybe we could do that too?